torch/csrc/api/include/torch/nn/parallel/data_parallel.h

```diff
@@ -0,0 +1,161 @@
+#pragma once
+
+#include <torch/cuda.h>
+#include <torch/nn/module.h>
+#include <torch/nn/pimpl.h>
+#include <torch/tensor.h>
+
+#include <torch/csrc/autograd/functions/comm.h>
+#include <torch/csrc/cuda/comm.h>
+
+#include <ATen/Device.h>
+#include <ATen/Error.h>
+#include <ATen/OptionsGuard.h>
+#include <ATen/Parallel.h>
+#include <ATen/TensorOptions.h>
+#include <ATen/optional.h>
+
+#include <cstddef>
+#include <exception>
+#include <iostream>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+namespace torch {
+namespace nn {
+namespace parallel {
+
+/// Replicates a module on the given list of devices.
+/// A replica is created by calling `clone()` on the module. For this, the
+/// module must inherit from `nn::Cloneable`, or define its own `clone()`
+/// method, which is expected to perform a deep copy of the module.
+template <typename ModuleType>
+std::vector<std::shared_ptr<ModuleType>> replicate(
+    const std::shared_ptr<ModuleType>& module,
+    const std::vector<Device>& devices) {
+  std::vector<std::shared_ptr<ModuleType>> replicas;
+  replicas.reserve(devices.size());
+  for (const auto& device : devices) {
+    // Here we rely on the property tensors are never (or should never be)
+    // allocated on any particular device, but always the default device, e.g.
+    // in `torch::ones({3, 4})`, the device is unspecified and pulled from the
+    // current thread local default options. As such, we can here modify these
+    // thread local default options and thereby cause all tensors in the cloned
+    // module to be constructed directly on the device we want.
+    OptionsGuard guard(device);
+    replicas.push_back(std::static_pointer_cast<ModuleType>(module->clone()));
```

Comment:
Hmm that's a bit surprising. I'd expect `clone()` to completely ignore the thread local device and just put them wherever the original params where. This way makes sense too I guess 😕